zkevm-circuits/src/util/cell_manager.rs

```diff
@@ -1,8 +1,269 @@
+    ) -> Result<AssignedCell<F, F>, Error> {
+        region.assign_advice(
+            || {
+                format!(
+                    "Cell column: {:?} and rotation: {}",
+                    self.column, self.rotation
+                )
+            },
+            self.column,
+            offset + self.rotation,
+            || value,
+        )
+    }
+}
+
+impl<F: Field> Expr<F> for Cell<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+}
+
+impl<F: Field> Expr<F> for &Cell<F> {
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+}
+
+#[derive(Debug, Clone)]
+/// CellColumn
+pub(crate) struct CellColumn {
+    pub advice: Column<Advice>,
+    pub cell_type: CellType,
+    pub idx: usize,
+}
+
+impl CellColumn {
+    pub fn new(advice: Column<Advice>, cell_type: CellType, idx: usize) -> CellColumn {
+        CellColumn {
+            advice,
+            cell_type,
+            idx,
+        }
+    }
+
+    pub fn expr<F: Field>(&self, meta: &mut ConstraintSystem<F>) -> Expression<F> {
+        query_expression(meta, |meta| meta.query_advice(self.advice, Rotation::cur()))
+    }
+}
+
+/// CellManagerStrategy
+pub(crate) trait CellManagerStrategy {
+    type Stats;
+
+    /// on_creation
+    fn on_creation(&mut self, columns: &mut CellManagerColumns);
+    /// query_cell
+    fn query_cell<F: Field>(
+        &mut self,
+        columns: &mut CellManagerColumns,
+        meta: &mut ConstraintSystem<F>,
+        cell_type: CellType,
+    ) -> Cell<F>;
+    /// get_height
     fn get_height(&self) -> usize;
+
+    fn get_stats(&self, columns: &CellManagerColumns) -> Self::Stats;
+}
+
+#[derive(Default, Debug, Clone)]
+pub(crate) struct CellManagerColumns {
+    columns: HashMap<CellType, Vec<CellColumn>>,
+    columns_list: Vec<CellColumn>,
+}
+
+impl CellManagerColumns {
+    pub fn add_column(&mut self, cell_type: CellType, column: Column<Advice>) {
+        let idx = self.columns_list.len();
+        let cell_column = CellColumn::new(column, cell_type, idx);
+        self.columns_list.push(cell_column.clone());
+        if let Some(columns) = self.columns.get_mut(&cell_type) {
+            columns.push(cell_column);
+        } else {
+            self.columns.insert(cell_type, vec![cell_column]);
+        }
+    }
+
+    pub fn get_cell_type_width(&self, cell_type: CellType) -> usize {
+        if let Some(columns) = self.columns.get(&cell_type) {
+            columns.len()
+        } else {
+            0
+        }
+    }
+
+    pub fn get_column(&self, cell_type: CellType, column_idx: usize) -> Option<&CellColumn> {
+        if let Some(columns) = self.columns.get(&cell_type) {
+            columns.get(column_idx)
+        } else {
+            None
+        }
+    }
+
+    pub fn columns(&self) -> Vec<CellColumn> {
+        self.columns_list.clone()
+    }
+
+    pub fn get_width(&self) -> usize {
+        self.columns.values().fold(0, |acc, g| acc + g.len())
```

Comment:
```suggestion
        self.columns_list.len()
```
This should be the same right?

Reply:
Yes, at the beggining there were no colums_list (btw, if you think of a better name...)